test/validation/api/dma/dma.c

```diff
@@ -1603,6 +1796,211 @@ static void test_dma_pkt_to_pkt_multiple_offsets(void)
+	setup_segs(dst_seg, pkts[dst_idx], seg_len, num_dst_seg, opts->seg_alloc);
+
+	odp_dma_transfer_param_init(&trs_param);
+	trs_param.src_format = ODP_DMA_FORMAT_PACKET;
+	trs_param.dst_format = ODP_DMA_FORMAT_PACKET;
+	trs_param.num_src = num_src_seg;
+	trs_param.num_dst = num_dst_seg;
+	trs_param.src_seg = src_seg;
+	trs_param.dst_seg = dst_seg;
+	trs_param.opts.seg_free = opts->seg_free;
+	trs_param.opts.unique_src_segs = opts->unique_src_segs;
+	trs_param.opts.single_pool = opts->single_pool;
+	trs_param.opts.seg_alloc = opts->seg_alloc;
+	trs_param.opts.unique_dst_segs = opts->unique_dst_segs;
+
+	if (opts->seg_alloc)
+		trs_param.dst_seg_pool = global.pkt_pool;
+
+	if (mask == ODP_DMA_COMPL_SYNC)
+		ret = do_transfer(dma, &trs_param, 0, 1);
+	else
+		ret = do_transfer_async(dma, &trs_param, mask, 0);
+
+	if (ret == 1) {
+		CU_ASSERT(check_equal(odp_packet_data(pkts[src_idx]),
+				      odp_packet_data(pkts[dst_idx]), seg_len) == 0);
+
+		if (opts->seg_free) {
+			/* The transfer-freed packet should be available now. */
+			pkt = odp_packet_alloc(pool, global.pkt_len);
+
+			CU_ASSERT(pkt != ODP_PACKET_INVALID);
+
+			pkts[src_idx] = pkt;
+		}
+	}
+
+	if (pkts[src_idx] != ODP_PACKET_INVALID)
+		odp_packet_free(pkts[src_idx]);
+
+	odp_packet_free(pkts[dst_idx]);
+
+	if (opts->seg_free)
+		CU_ASSERT(odp_pool_destroy(pool) == 0);
+
+	CU_ASSERT(odp_dma_destroy(dma) == 0);
+}
+
+static void test_dma_src_seg_free_sync(void)
+{
+	const opts_t opts = { .seg_free = true,
+			      .unique_src_segs = true,
+			      .single_pool = false,
+			      .seg_alloc = false,
+			      .unique_dst_segs = true };
+
+	test_dma_pkt_to_pkt_opts(ODP_DMA_COMPL_SYNC, &opts);
+}
+
+static void test_dma_src_seg_free_single_pool_sync(void)
+{
+	const opts_t opts = { .seg_free = true,
+			      .unique_src_segs = true,
+			      .single_pool = true,
+			      .seg_alloc = false,
+			      .unique_dst_segs = true };
+
+	test_dma_pkt_to_pkt_opts(ODP_DMA_COMPL_SYNC, &opts);
+}
+
+static void test_dma_src_seg_free_multiple_offsets_sync(void)
+{
+	const opts_t opts = { .seg_free = true,
+			      .unique_src_segs = false,
+			      .single_pool = false,
+			      .seg_alloc = false,
+			      .unique_dst_segs = true };
+
+	test_dma_pkt_to_pkt_opts(ODP_DMA_COMPL_SYNC, &opts);
+}
+
+static void test_dma_dst_seg_alloc_sync(void)
+{
+	const opts_t opts = { .seg_free = false,
+			      .unique_src_segs = true,
+			      .single_pool = false,
+			      .seg_alloc = true,
+			      .unique_dst_segs = true };
+
+	test_dma_pkt_to_pkt_opts(ODP_DMA_COMPL_SYNC, &opts);
+}
+
+static void test_dma_dst_seg_alloc_multiple_offsets_sync(void)
+{
+	const opts_t opts = { .seg_free = false,
+			      .unique_src_segs = true,
+			      .single_pool = false,
+			      .seg_alloc = true,
+			      .unique_dst_segs = false };
+
+	test_dma_pkt_to_pkt_opts(ODP_DMA_COMPL_SYNC, &opts);
+}
+
+static void test_dma_src_seg_free_dst_seg_alloc(void)
```

Comment:
`test_dma_src_seg_free_dst_seg_alloc_sync` for consistency?

Reply:
Idea was that as there are no other variants (poll/event) for this function, there's no "sync" here (similar to `test_dma_pkt_to_pkt_multiple_offsets()`).